core/src/apps/cardano/certificates.py

```diff
@@ -0,0 +1,272 @@
+from trezor.messages import CardanoCertificateType, CardanoPoolRelayType
+
+from apps.common import cbor
+
+from .address import (
+    get_address_bytes_unsafe,
+    get_public_key_hash,
+    validate_reward_address,
+)
+from .helpers import INVALID_CERTIFICATE
+from .helpers.constants import LOVELACE_MAX_SUPPLY
+from .helpers.paths import SCHEMA_STAKING
+
+if False:
+    from trezor.messages.CardanoTxCertificateType import CardanoTxCertificateType
+    from trezor.messages.CardanoPoolParametersType import CardanoPoolParametersType
+    from trezor.messages.CardanoPoolRelayParametersType import (
+        CardanoPoolRelayParametersType,
+    )
+    from trezor.messages.CardanoPoolOwnerType import CardanoPoolOwnerType
+    from trezor.messages.CardanoPoolMetadataType import CardanoPoolMetadataType
+    from typing import List, Tuple, Optional
+    from . import seed
+
+
+POOL_HASH_SIZE = 28
+VRF_KEY_HASH_SIZE = 32
+POOL_METADATA_HASH_SIZE = 32
+PUBLIC_KEY_HASH_SIZE = 28
+IPV4_ADDRESS_SIZE = 4
+IPV6_ADDRESS_SIZE = 16
+
+MAX_URL_LENGTH = 64
+MAX_PORT_NUMBER = 65535
+
+
+def validate_certificate(
+    certificate: CardanoTxCertificateType, protocol_magic: int, network_id: int
+) -> None:
+    if certificate.type in [
+        CardanoCertificateType.STAKE_DELEGATION,
+        CardanoCertificateType.STAKE_REGISTRATION,
+        CardanoCertificateType.STAKE_DEREGISTRATION,
+    ]:
+        if not SCHEMA_STAKING.match(certificate.path):
+            raise INVALID_CERTIFICATE
+
+    if certificate.type == CardanoCertificateType.STAKE_DELEGATION:
+        if not certificate.pool or len(certificate.pool) != POOL_HASH_SIZE:
+            raise INVALID_CERTIFICATE
+
+    if certificate.type == CardanoCertificateType.STAKE_POOL_REGISTRATION:
+        _validate_pool_parameters(
+            certificate.pool_parameters, protocol_magic, network_id
+        )
+
+
+def cborize_certificate(
+    keychain: seed.Keychain, certificate: CardanoTxCertificateType
+) -> Tuple:
```

Comment:
I noticed the cardano and many other coins (pretty much all except Bitcoin) are missing from mypy targets and now that I added Cardano there, there us quite a lot of type errors (>100) that are in master as well. Other coins left out from mypy target have the same problem, e.g. Binance coin. Given the amount and of errors and their span, I just focused on fixing the ones related to certificates.

Why are most coins missing from mypy targets in the Makefile?

Reply:
we started using type annotations before we started enforcing typing via mypy, so in most of the code the type annotations were never actually in use, and consequently there is a lot of typing errors. We plan to fix typing and add targets gradually. If you're able to get mypy to pass for certificates.py, please add that single file to the mypy target; if not, don't worry about it for now.

Of course it would be nice to get typing right for the whole Cardano app at some point, and having a bunch of files that typecheck is a good start in that direction.

Reply:
I tried adding just the certificates.py file and even though the file itself does not fail mypy checks, mypy enters also the files that are being imported in certificates.py which results in other approximately 30 errors which would bloat this PR. Alternatively, I can disable type checks for each of those files, but this can complicate future efforts to make the files pass mypy checks as it would be silencing type errors related to them. So I would rather leave it as it is right now